docs/rfc/0000-import-from-images.md

```diff
@@ -0,0 +1,321 @@
+without sacrificing repeatability, and introduces a low-stakes way to
+trial using container images for packaging.
+
+## Design
+
+### User interface
+
+    jk run --lib jkcfg/kubernetes:0.2.1 -I ../mycharts generate.js
+
+`--lib` is put forward here as small and guessable. `--image` is an
+alternative, but that doesn't indicate it's a _library_ rather than
+the thing to execute.
+
+### Downloading and caching images
+
+When a library image is required for excuting a `jk` script, these
+things shall happen:
+
+ 1. The image ref is resolved to a path within the cache directory
+
+    1.2. whether the ref is a tag or a digest, the path is expected to
+         be a symlink to a blob which is the manifest. Note that the
+         digest ref for an image is not the digest of the manifest
+         blob (I don't know what it is, maybe the digest of the
+         gzipped file?)
+
+ 2. If the file does not exist, it is resolved using OCI distribution,
+    and written into the cache (as per the expectations in the
+    previous step)
+
+ 3. The file at the path, or linked, is a manifest (MIME type
+    application/vnd.oci.image.manifest.v1+json); if necessary, it's
+    converted to that from a Docker image manifest
+
+    3.1. If the fetched file is a manifest list, the appropriate
+         manifest from it is fetched in turn.
+
+ 4. For each layer in the image (we don't care about the config,
+    that's for running the image), if it is missing then fetch it,
+    verify its digest, and put it in the cache blob store
+
+ 5. A filesystem based on the image layers is added to the module
+    search path.
+
+### Layout of images
+
+Container images include a whole filesystem. Where should `jk` look
+for modules?
+
+ * we will want image layers to compose so they can be remuxed, so it
+   makes sense to put modules in the same place every time
+ * e.g., just chuck everything under `/jk/modules/`, so if you jam a
+   bunch of layers together in an image, it'll just look like a
+   directory with all the libraries under `/jk/modules/`.
+
+### Using images in the module search path
+
+In two steps:
+
+ - a. adapt the module resolvers to allow filesystems represented by
+   http.FileSystem (or similar)
+ - b. write a union filesystem that understands the diff layers of an OCI
+   image
+
+Since modules may be in some abstraction of a filesystem, access to
+resources will need to go through the abstraction too, as will any
+other built-in code that loads files (like schema validation).
+
+The work for a.) was done in [PR 307](#307). The remainder, i.e., b.)
+is done in [PR 315][#315].
+
+#### Building images
+
+Although this RFC is not about tooling for building images, it's worth
+mentioning here how one would go about it.
+
+Using a Dockerfile, you can make an archive of your current
+node_modules contents:
+
+```
+$ cat > Dockerfile <<EOF
+FROM scratch
+WORKINGDIR /jk
+# NB COPY copies the _contents_ of directories
+COPY node_modules ./
```

Comment:
I think it's a good idea to specify the actual path in the image the modules are expected to be. Above, you give the `/jk/modules` example but it seems it's `/jk/node_modules` this docker image.

Reply:
Oh good point, I need to adjust this.

Reply:
COPY copies the contents of a directory, so with the workingdir set to /jk/modules, this should copy everything in `node_modules` to /jk/modules/; i.e., act like

    cp node_modules/* /jk/modules/

Reply:
.. and of course it's WORKDIR not WORKINGDIR